### ISSUE TITLE
Having a couple problems with jumping over obstacles

### ISSUE THREAD

**Issue by Skaruts** [alexjhetherington/godot-character-controller-example#5]
(My version of the demo project is attached below.)

I found two issues with jumping over things, which may be related to the stepping code. I added some boxes to the test map. My jump code is fairly simple, just add `jump_force` to the `desired_vertical_velocity.y` when space is pressed. 

Jumping mostly works fine, except that when hugging an obstacle, pressing forward into it and jumping:

- there's a bit of a bump in the movement, when the player reaches the height of the orange boxes (not very noticeable in the gif)
- the player will stick to the red box as if it was a stair step. As in, the jump will be cut short, at that height. 

![fps_controller_jump_bug_2](https://github.com/alexjhetherington/godot-character-controller-example/assets/6277234/fb56d107-5472-411f-ae9c-e41d1a291954)

The second issue may disappear by changing the `step_height` or the `jump_force`. It seems to only happen under some circumstances, and with the current settings the red box is the only one with the right height for it to happen. But the first issue seems to be always there, except on smaller obstacles, like the white box.

Here's the demo project with jumping. It's a bit bigger because it already includes Jolt, and TBLoader, another test map and a few textures.
[godot-character-controller-example-main.zip](https://github.com/alexjhetherington/godot-character-controller-example/files/13256437/godot-character-controller-example-main.zip)

The few changes I made to the code all have the comment `# (skaruts)`, to make them easier to find, if needed.

**Comment by alexjhetherington** [comment]
Introducing jumping invites many decisions about how a character controller should work that are game specific (or at the least opinion driven). I don't want to be directly involved with those discussions in the scope of keeping this repo up-to-date, which is why there is no jumping included, and why I won't make any changes to the repo based on this issue (even though arguably there is a bug, see below).

But I will gladly offer my general thoughts and advice.

**What's happening**

"Usual" Behaviour:

- The player is stepping up onto *all of the blocks*
- The two cases you indicated are more noticeable

First case:

- As the player jumps, their upward velocity slowly reduces. When the player is almost at the peak of their jump, the step algorithm detects a surface it can step up; and you see a "bump." It's noticeable because the player had almost stopped vertical movement, then they suddenly get some more.

Second case:
- Near the start of the jump, the step algorithm detected a step and took a step
- At this particular block height and jump speed, the step algorithm collided _right on the edge of the block_; where an unexpected surface normal is reported i.e. the algorithm thinks "I can step up onto a walkable slope"; which sets the state to grounded, zero-ing vertical velocity
- I'll comment on implications below

**Discussion for solving**

It depends on what you want movement behaviour to be!

If you want to disallow stepping while in midair, you can disable the stepping algorithm while in midair! Find the line where `do_step = true` and add a condition `and grounded` to the if statement above. No more stepping and your problems will be gone, I think.

There is an argument that stepping while in midair is good actually. You can see me make this argument in my own devlog: https://www.youtube.com/watch?v=N-vXN4Nnk9E&t=1216s (17:04). One way you could implement that is by stepping (though there are other ways).

If you wanted to keep mid-air stepping let's assume stepping onto blocks keeps your vertical momentum, and so we need to address the strange behaviour in your second case. There are many complicated edge cases. What I will say for now is (it requires understanding the stepping algorithm, and what exactly is happening in this bug case):

1. A collision that happens on an edge is bound to return a normal that represents a "slope"
2. It's surprising to me that a cylinder moving laterally can *possibly* collide with a horizontal edge at the top of a vertical surface; but physics engines are complicated and floating point rounding means anything can happen :D 
3. Turning up `safe_margin` (e.g. to `0.01`) eliminates the issue in this case by more aggressively separating the player from any colliding surface before each movement
4. This is exactly the kind of thing that `safe_margin` is there for; it's a common technique; we must acknowledge that tweaking this value is not particularly scientific and might break other edge cases: https://github.com/alexjhetherington/godot-character-controller-example/issues/4

Hope that all makes sense. 

tl:dr - disable stepping while mid-air and you'll be fine

Closing the issue without changing for now but I might disable mid-air stepping in future if more people see this issue.

Feel free to keep discussing here tho

**Comment by alexjhetherington** [comment]
(if you wanted mid-air stepping, you could also just force grounded = false in the case of a mid-air stair step, but I wanted an excuse to talk about safe_margin that I could point people to)

**Comment by Skaruts** [comment]
Yup, adding the `grounded` condition seems to have fixed the issue.

~~Out of curiosity, what would mid-air stepping be useful for? I can't think of anything.~~ Wait, I forgot you linked to a video talking about it. Gonna watch it.
